tsdb/series_collection.go

```diff
@@ -0,0 +1,334 @@
+package tsdb
+
+import (
+	"sync"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/influxdata/influxdb/pkg/bytesutil"
+
+	"github.com/influxdata/influxdb/models"
+)
+
+// SeriesCollection is a struct of arrays representation of a collection of series that allows
+// for efficient filtering.
+type SeriesCollection struct {
+	Points     []models.Point
+	Keys       [][]byte
+	SeriesKeys [][]byte
+	Names      [][]byte
+	Tags       []models.Tags
+	Types      []models.FieldType
+	SeriesIDs  []SeriesID
+
+	// Keeps track of invalid entries.
+	Dropped     uint64
+	DroppedKeys [][]byte
+	Reason      string
+
+	// Used by the concurrent iterators to stage drops. Inefficient, but should be
+	// very infrequently used.
+	state *seriesCollectionState
+}
+
+// seriesCollectionState keeps track of concurrent iterator state.
+type seriesCollectionState struct {
+	mu     sync.Mutex
+	reason string
+	index  map[int]struct{}
+}
+
+// NewSeriesCollection builds a SeriesCollection from a slice of points. It does some filtering
+// of invalid points.
+func NewSeriesCollection(points []models.Point) *SeriesCollection {
+	out := &SeriesCollection{
+		Points: append([]models.Point(nil), points...),
+		Keys:   make([][]byte, 0, len(points)),
+		Names:  make([][]byte, 0, len(points)),
+		Tags:   make([]models.Tags, 0, len(points)),
+		Types:  make([]models.FieldType, 0, len(points)),
+	}
+
+	for _, pt := range points {
+		out.Keys = append(out.Keys, pt.Key())
+		out.Names = append(out.Names, pt.Name())
+		out.Tags = append(out.Tags, pt.Tags())
+
+		fi := pt.FieldIterator()
+		fi.Next()
+		out.Types = append(out.Types, fi.Type())
+	}
+
+	return out
+}
+
+// Duplicate returns a copy of the SeriesCollection. The slices are shared. Appending to any of
+// them may or may not be reflected.
+func (s SeriesCollection) Duplicate() *SeriesCollection { return &s }
+
+// Length returns the length of the first non-nil slice in the collection, or 0 if there is no
+// non-nil slice.
+func (s *SeriesCollection) Length() int {
+	switch {
+	case s.Points != nil:
+		return len(s.Points)
+	case s.Keys != nil:
+		return len(s.Keys)
+	case s.SeriesKeys != nil:
+		return len(s.SeriesKeys)
+	case s.Names != nil:
+		return len(s.Names)
+	case s.Tags != nil:
+		return len(s.Tags)
+	case s.Types != nil:
+		return len(s.Types)
+	case s.SeriesIDs != nil:
+		return len(s.SeriesIDs)
+	default:
+		return 0
+	}
+}
+
+// Copy will copy the element at src into dst in all slices that can: x[dst] = x[src].
+func (s *SeriesCollection) Copy(dst, src int) {
+	if dst == src {
+		return
+	}
+	udst, usrc := uint(dst), uint(src)
+	if len := uint(len(s.Points)); udst < len && usrc < len {
+		s.Points[udst] = s.Points[usrc]
+	}
```

Comment:
I'd personally rename `len` to `n` so you don't shadow the `len()` function. It makes it a little easier to read, however, it's not really a big deal.